test/adapters/level_zero/event_cache_tests.cpp

```diff
@@ -0,0 +1,170 @@
+#include "uur/fixtures.h"
+#include "uur/raii.h"
+
+#include <map>
+#include <string>
+
+template <typename... Args> auto combineFlags(std::tuple<Args...> tuple) {
+    return std::apply([](auto... args) { return (... |= args); }, tuple);
+}
+
+extern std::map<std::string, int> *ZeCallCount;
+
+using FlagsTupleType = std::tuple<ur_queue_flags_t, ur_queue_flags_t,
+                                  ur_queue_flags_t, ur_queue_flags_t>;
+
+#define ASSERT_SUCCESS_OR_UNSUPPORTED(ret)                                     \
+    if (ret == UR_RESULT_ERROR_UNSUPPORTED_FEATURE) {                          \
+        GTEST_SKIP();                                                          \
+    } else {                                                                   \
+        ASSERT_EQ(ret, UR_RESULT_SUCCESS);                                     \
+    }
+
+struct urEventCacheTest : uur::urContextTestWithParam<FlagsTupleType> {
+    void SetUp() override {
+        UUR_RETURN_ON_FATAL_FAILURE(urContextTestWithParam::SetUp());
+
+        flags = combineFlags(getParam());
+
+        ur_queue_properties_t props;
+        props.flags = flags;
+        ASSERT_SUCCESS(urQueueCreate(context, device, &props, &queue));
+        ASSERT_NE(queue, nullptr);
+
+        ASSERT_SUCCESS(urMemBufferCreate(context, UR_MEM_FLAG_WRITE_ONLY, size,
+                                         nullptr, &buffer));
+
+        (*ZeCallCount)["zeEventCreate"] = 0;
+        (*ZeCallCount)["zeEventDestroy"] = 0;
+    }
+
+    void TearDown() override {
+        if (buffer) {
+            EXPECT_SUCCESS(urMemRelease(buffer));
+        }
+        if (queue) {
+            ASSERT_SUCCESS_OR_UNSUPPORTED(urQueueRelease(queue));
+        }
+        UUR_RETURN_ON_FATAL_FAILURE(urContextTestWithParam::TearDown());
+    }
+
+    auto enqueueWork(ur_event_handle_t *hEvent, int data) {
+        input.assign(count, data);
+        ASSERT_SUCCESS_OR_UNSUPPORTED(urEnqueueMemBufferWrite(
+            queue, buffer, false, 0, size, input.data(), 0, nullptr, hEvent));
+    }
+
+    void verifyData() {
+        std::vector<uint32_t> output(count, 1);
+        ASSERT_SUCCESS_OR_UNSUPPORTED(urEnqueueMemBufferRead(
+            queue, buffer, true, 0, size, output.data(), 0, nullptr, nullptr));
+
+        if (!(flags & UR_QUEUE_FLAG_OUT_OF_ORDER_EXEC_MODE_ENABLE)) {
+            ASSERT_EQ(input, output);
+        }
+    }
+
+    const size_t count = 1024;
+    const size_t size = sizeof(uint32_t) * count;
+    ur_mem_handle_t buffer = nullptr;
+    ur_queue_handle_t queue = nullptr;
+    std::vector<uint32_t> input;
+    ur_queue_flags_t flags;
+};
+
+TEST_P(urEventCacheTest, eventsReuseNoVisibleEvent) {
+    static constexpr int numIters = 16;
+    static constexpr int numEnqueues = 128;
+
+    for (int i = 0; i < numIters; i++) {
+        for (int j = 0; j < numEnqueues; j++) {
+            enqueueWork(nullptr, i * numEnqueues + j);
+        }
+        ASSERT_SUCCESS_OR_UNSUPPORTED(urQueueFinish(queue));
```

Comment:
it's odd that the UNSUPPORTED is returned by urQueueFinish. Did you investigate what specifically is returning that?